# **EDA**
El proceso que haremos será: 

1.   Verificacion de datos iniciales
2.   Datos nulos y datos duplicados
3.   Outliers

In [1]:
#Corremos librerias
%run ./START.ipynb

In [2]:
#Visualizamos el head del dataset
df.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008805           M            Y               Y             0   
2  5008806           M            Y               Y             0   
3  5008808           F            N               Y             0   
4  5008809           F            N               Y             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0            427500               Working               Higher education   
1            427500               Working               Higher education   
2            112500               Working  Secondary / secondary special   
3            270000  Commercial associate  Secondary / secondary special   
4            270000  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  ...  AGE  DAYS_EMPLOYED  \
0        Civil marriage   Rented apartment  ...   33          -4542   
1        Civil marriage   Rented apartment  ...   33          -4542   
2               Married  House / apartment  ...   59          -1134   
3  Single / not married  House / apartment  ...   52          -3051   
4  Single / not married  House / apartment  ...   52          -3051   

   YEARS_EMPLOYED  FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL  \
0              12           1                1           0           0   
1              12           1                1           0           0   
2               3           1                0           0           0   
3               8           1                0           1           1   
4               8           1                0           1           1   

   OCCUPATION_TYPE CNT_FAM_MEMBERS  APPROVED  
0              NaN               2         1  
1              NaN               2         1  
2   Security staff               2         1  
3      Sales staff               1         1  
4      Sales staff               1         1  

[5 rows x 21 columns]

In [3]:
df.shape

(438557, 21)

In [4]:
#Nos aseguramos que todos los datos NULL hayan quedado efectivamente borrados con una limpieza más
df_clean = df.mask(df == 'NULL').dropna()
df_clean.shape

(304354, 21)

De los 438557 luego de la limpieza, nos quedamos con 304354 datos.

In [5]:
#Nos percatamos que en la variable CNT_CHILDREN hay datos atipicos que no agregan información de valor a nuestro estudio, procedemos a obviarlos
var=df_clean[df_clean["CNT_CHILDREN"]>=6].index
df_cleaner=df_clean.drop(var)
df_cleaner.shape

(304332, 21)

df_cleaner es un dataset "mas limpio" donde nos quedan 304332 filas habiendo realizado una limpieza por cantidad de hijos mayor a 6

In [6]:
df_cleaner.duplicated().sum()

0

Comprobamos que la cantidad de valores en nuestro set de datos no cambia, por lo cual podemos asumir que no hay datos duplicados

In [7]:
#realizamos un describe para tener una idea general de los datos que trae el dataframe hasta el momento
df_cleaner.describe(include='all').T

count unique                            top    freq  \
ID                   304332.0    NaN                            NaN     NaN   
CODE_GENDER            304332      2                              F  190405   
FLAG_OWN_CAR           304332      2                              N  178230   
FLAG_OWN_REALTY        304332      2                              Y  207655   
CNT_CHILDREN         304332.0    NaN                            NaN     NaN   
AMT_INCOME_TOTAL     304332.0    NaN                            NaN     NaN   
NAME_INCOME_TYPE       304332      5                        Working  190207   
NAME_EDUCATION_TYPE    304332      5  Secondary / secondary special  204770   
NAME_FAMILY_STATUS     304332      5                        Married  212708   
NAME_HOUSING_TYPE      304332      6              House / apartment  269995   
DAYS_BIRTH           304332.0    NaN                            NaN     NaN   
AGE                  304332.0    NaN                            NaN     NaN   
DAYS_EMPLOYED        304332.0    NaN                            NaN     NaN   
YEARS_EMPLOYED       304332.0    NaN                            NaN     NaN   
FLAG_MOBIL           304332.0    NaN                            NaN     NaN   
FLAG_WORK_PHONE      304332.0    NaN                            NaN     NaN   
FLAG_PHONE           304332.0    NaN                            NaN     NaN   
FLAG_EMAIL           304332.0    NaN                            NaN     NaN   
OCCUPATION_TYPE        304332     18                       Laborers   78240   
CNT_FAM_MEMBERS      304332.0    NaN                            NaN     NaN   
APPROVED             304332.0    NaN                            NaN     NaN   

                               mean            std        min         25%  \
ID                   6022997.399012  569812.051433  5008806.0  5617738.75   
CODE_GENDER                     NaN            NaN        NaN         NaN   
FLAG_OWN_CAR                    NaN            NaN        NaN         NaN   
FLAG_OWN_REALTY                 NaN            NaN        NaN         NaN   
CNT_CHILDREN               0.510301       0.760888        0.0         0.0   
AMT_INCOME_TOTAL      194878.890685  117005.683616    27000.0    135000.0   
NAME_INCOME_TYPE                NaN            NaN        NaN         NaN   
NAME_EDUCATION_TYPE             NaN            NaN        NaN         NaN   
NAME_FAMILY_STATUS              NaN            NaN        NaN         NaN   
NAME_HOUSING_TYPE               NaN            NaN        NaN         NaN   
DAYS_BIRTH            -14806.741039     3487.30126   -24770.0    -17439.0   
AGE                        40.57042        9.56118       21.0        33.0   
DAYS_EMPLOYED          -2621.946877    2400.307017   -17531.0     -3508.0   
YEARS_EMPLOYED             7.193979       6.588208        0.0         3.0   
FLAG_MOBIL                      1.0            0.0        1.0         1.0   
FLAG_WORK_PHONE            0.248143       0.431936        0.0         0.0   
FLAG_PHONE                  0.28557       0.451686        0.0         0.0   
FLAG_EMAIL                 0.117543       0.322066        0.0         0.0   
OCCUPATION_TYPE                 NaN            NaN        NaN         NaN   
CNT_FAM_MEMBERS            2.297307       0.922294        1.0         2.0   
APPROVED                   0.840198       0.366423        0.0         1.0   

                           50%         75%        max  
ID                   6047835.5  6449034.25  7999952.0  
CODE_GENDER                NaN         NaN        NaN  
FLAG_OWN_CAR               NaN         NaN        NaN  
FLAG_OWN_REALTY            NaN         NaN        NaN  
CNT_CHILDREN               0.0         1.0        5.0  
AMT_INCOME_TOTAL      180000.0    225000.0  6750000.0  
NAME_INCOME_TYPE           NaN         NaN        NaN  
NAME_EDUCATION_TYPE        NaN         NaN        NaN  
NAME_FAMILY_STATUS         NaN         NaN        NaN  
NAME_HOUSING_TYPE          NaN

In [8]:
#Nos percatamos que la variable AMT_INCOME_TOTAL existen datos atipicos (outliers) que desbalancean nuestra muestra fuertemente, eliminaremos aquellos que superan un salario de 300.000 USD anuales y renombramos nuestro dataset a "new_df" 
var=df_cleaner[df_cleaner["AMT_INCOME_TOTAL"]>=300000].index
new_df=df_cleaner.drop(var)
new_df.shape

(268801, 21)

In [9]:
porcentaje_newdf=268801/438557
porcentaje_newdf

0.6129214674489291

Como resultado final de nuestro data wrangling obtuvimos un nuevo dataframe "new_df" con el 61,29% de los datos iniciales con los que consideramos que podremos trabajar el modelo más eficientemente.

El procedimiento será hacer una tabla estilo dataframe con todas las columnas donde vamos a obtener:

*   Cantidad de Datos
*   Tipo de dato
*   Valores únicos
*   Valores nulos
*   Promedio
*   Máximo
*   Mínimo
*   Quartiles
*   Sesgo y kurtosis

Con toda esta información podremos obtener cierta relación donde determinemos variables numéricas o categóricas, cuales tienen mucha variabilidad, promedio y medidas centrales.

In [10]:
from pandas.api.types import is_numeric_dtype

def info_univariado(new_df):
  df_info = pd.DataFrame(columns=['Cantidad', 'Tipo' , 'Missing', 'Unicos', 'Numeric'])
  for col in new_df:
    data_series = new_df[col]
    df_info.loc[col] = [data_series.count(), data_series.dtype, data_series.isnull().sum(), data_series.nunique(), is_numeric_dtype(data_series)]
    df_describe = new_df.describe(include='all').T[['top', 'mean', 'std', 'min', '25%', '50%', '75%', 'max']]
    df_stats = pd.DataFrame([new_df.skew(numeric_only=True), new_df.kurtosis(numeric_only=True)], index=['sesgo', 'kurt']).T
  return pd.concat([df_info,pd.concat([df_describe, df_stats], axis=1)], axis=1).fillna('0')

stats = info_univariado(new_df)
stats

Cantidad    Tipo  Missing  Unicos  Numeric  \
ID                     268801   int64        0  268784     True   
CODE_GENDER            268801  object        0       2    False   
FLAG_OWN_CAR           268801  object        0       2    False   
FLAG_OWN_REALTY        268801  object        0       2    False   
CNT_CHILDREN           268801   int64        0       6     True   
AMT_INCOME_TOTAL       268801   int64        0     492     True   
NAME_INCOME_TYPE       268801  object        0       5    False   
NAME_EDUCATION_TYPE    268801  object        0       5    False   
NAME_FAMILY_STATUS     268801  object        0       5    False   
NAME_HOUSING_TYPE      268801  object        0       6    False   
DAYS_BIRTH             268801   int64        0   13944     True   
AGE                    268801   int64        0      48     True   
DAYS_EMPLOYED          268801   int64        0    8716     True   
YEARS_EMPLOYED         268801   int64        0      48     True   
FLAG_MOBIL             268801   int64        0       1     True   
FLAG_WORK_PHONE        268801   int64        0       2     True   
FLAG_PHONE             268801   int64        0       2     True   
FLAG_EMAIL             268801   int64        0       2     True   
OCCUPATION_TYPE        268801  object        0      18    False   
CNT_FAM_MEMBERS        268801   int64        0       7     True   
APPROVED               268801   int64        0       2     True   

                                               top            mean  \
ID                                               0  6021593.650414   
CODE_GENDER                                      F               0   
FLAG_OWN_CAR                                     N               0   
FLAG_OWN_REALTY                                  Y               0   
CNT_CHILDREN                                     0        0.512561   
AMT_INCOME_TOTAL                                 0   166585.414169   
NAME_INCOME_TYPE                           Working               0   
NAME_EDUCATION_TYPE  Secondary / secondary special               0   
NAME_FAMILY_STATUS                         Married               0   
NAME_HOUSING_TYPE                House / apartment               0   
DAYS_BIRTH                                       0   -14735.897169   
AGE                                              0       40.376118   
DAYS_EMPLOYED                                    0    -2605.343845   
YEARS_EMPLOYED                                   0         7.14895   
FLAG_MOBIL                                       0             1.0   
FLAG_WORK_PHONE                                  0        0.254672   
FLAG_PHONE                                       0        0.283358   
FLAG_EMAIL                                       0        0.108232   
OCCUPATION_TYPE                           Laborers               0   
CNT_FAM_MEMBERS                                  0        2.300497   
APPROVED                                         0        0.823978   

                               std        min        25%        50%  \
ID                   570099.995841  5008806.0  5609687.0  6047494.0   
CODE_GENDER                      0          0          0          0   
FLAG_OWN_CAR                     0          0          0          0   
FLAG_OWN_REALTY                  0          0          0          0   
CNT_CHILDREN              0.762822        0.0        0.0        0.0   
AMT_INCOME_TOTAL      57214.764259    27000.0   121500.0   157500.0   
NAME_INCOME_TYPE                 0          0          0          0   
NAME_EDUCATION_TYPE              0          0          0          0   
NAME_FAMILY_STATUS               0          0          0          0   
NAME_HOUSING_TYPE                0          0          0          0   
DAYS_BIRTH             3501.638776   -24770.0   -17390.0   -14494.0   
AGE                       9.601768       21.0       32.0       40.0   
DAYS_EMPLOYED          2410.767723   -17531.0    -3488.0    -1885.0   
YEARS_EMPLO

 **Análisis:**

* La variable "OCCUPATION_TYPE" tenía el 30,6% de valores missings antes de la limpieza de datos, y ahora cuenta con 0% valores missings.
* La variable "CODE_GENDER" nos da información sobre que la población a analizar en su mayoría es femenina
* NAME_INCOME_TYPE es una variable categórica que contiene 5 tipos de datos siendo el más frecuente "Working" es decir que la mayoría de personas obtienen sus ingresos ejerciendo algún tipo de trabajo
* La variable "NAME_HOUSING_TYPE" nos da información de que la mayoria de la población a analizar vive en una casa o apartamento. 
* De acuerdo con la variable NAME_FAMILY_STATUS contamos con 5 tipos de datos diferentes siendo el más frecuente "Married" (Casado).
* De acuerdo con la variable "FLAG_OWN_CAR" detectamos que la mayoría no posee vehículo
* La variable "NAME_HOUSING_TYPE" nos da información de que la mayoria de la población a analizar vive en una casa o apartamento. 
* De acuerdo con la variable NAME_FAMILY_STATUS contamos con 5 tipos de datos diferentes siendo el más frecuente "Married" (Casado).
* La variable "FLAG_OWN_CAR" permite observar que la mayoría de la población no posee vehículo
De las 3 variables escogidas, ninguna de ellas tiene datos missing lo que hace ideal a la hora de analizarlos
* La variable "FLAG_MOBIL" nos da información de que toda la población a analizar posee teléfono móvil. 
* De acuerdo con la variable "FLAG_PHONE" podemos detectar que la mayor parte de la población a analizar no posee teléfono fijo.
* En la variable "FLAG_EMAIL" detectamos que la mayoría de personas a analizar en nuestro dataset no poseen correo electronico. Esto nos da pie para poder buscar la mejor vía de contacto para cada caaso de scoring.